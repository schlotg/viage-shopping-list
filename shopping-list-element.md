# Viage Shopping List Tutorial

## Adding a Shopping List Element Component
