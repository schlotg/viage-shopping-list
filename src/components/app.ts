import { ShoppingList } from './shopping-list';
import { ShoppingListAdd } from './shopping-list-add';
import { createRouter, Component } from 'viage';

export enum States {
  HOME = 'home',
  ADD = 'add',
  EDIT = 'edit'
};

export class App extends Component {

  private title = 'Shopping List';

  constructor() {
    super('app');
  }
  init() {
    document.querySelector('title').textContent = this.title;

    this.attach('page', true);
    this.setHTML(`
      <h1 style="text-align: center">${this.title}</h1>
      <div style="width: 100%; min-width: 654px" attach="portal"></div>
    `);

    const router = createRouter('main', this.attachments.portal, 'HASH');
    router.addStates([
      { name: States.HOME, component: ShoppingList,  type: 'DEFAULT' },
      { name: States.ADD, component: ShoppingListAdd,  type: 'NORMAL' },
      { name: States.EDIT, component: ShoppingListAdd,  type: 'NORMAL' },
    ]);
    // start off by going to the state the page was loaded on
    router.start();
    return this;
  }
}
