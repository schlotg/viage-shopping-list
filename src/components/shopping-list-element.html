<div class="not-mobile list-item">
  <input attach="enabled" type="checkbox" ${this.purchased ? "checked": ""} />
  <span style="width: 30px; display: inline-block">${this.quantity}</span>
  <span style="width: 20%; display: inline-block">${this.name}</span>
  <span style="width: 40%; display: inline-block">${this.description}</span>
  <button attach="delete" style="padding: 1px; float: right">Delete</button>
  <button attach="edit" style="padding: 1px; float: right; margin-right: 10px;">Edit</button>
</div>
<div attach="m_edit" class="mobile list-item">
  <input attach="m_enabled" type="checkbox" ${this.purchased ? "checked": ""} />
  <span style="width: 30px; display: inline-block">${this.quantity}</span>
  <span style="width: 50%; display: inline-block">${this.name}</span>
  <button attach="m_delete" style="padding: 1px; float: right">X</button>
  <div style="display: block; margin-top:20px;">
    <label style="font-weight:800">Description: </label>
    <span style="width: 40%;" >${this.description}</span>
  </div>
</div>
