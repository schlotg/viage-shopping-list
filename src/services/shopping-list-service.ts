import { Service } from 'viage';

export interface Item {
  name: string;
  description: string;
  quantity: number;
  enabled: boolean;
  _id?: string; // added by the service
}

class ShoppingListServiceSingleton extends Service {

  private list: Item[] = [];
  private count = 1;

  private generateId() {
    this.count += 1;
    return this.count.toString(16);
  }

  private findIndex(id: string): number {
    let i = 0;
    const list = this.list;
    for (; i < list.length; ++i) {
      if (list[i]._id === id) { break; }
    }
    return i;
  }

  constructor() {
    super();
    if (window.localStorage) {
      const json = window.localStorage.getItem('ShoppingList');
      if (json) {
        const obj = JSON.parse(json) || {};
        this.count = obj.count || 0;
        const list = obj.list || [];
        list.forEach((e: Item) => this.addItem(e));
      }
    }
  }

  addItem(item: Item) {
    item._id = item._id || this.generateId();
    this.list.push(item);
    this.save();
  }

  removeItem(id: string) {
    const i = this.findIndex(id);
    this.list.splice(i, 1);
    this.save();
  }

  getItem(id: string): Item {
    const i = this.findIndex(id);
    return this.list[i];
  }

  forEach(cb: (item: Item, i?: number) => void) {
    this.list.forEach(cb);
  }

  save() {
    if (window.localStorage) {
      window.localStorage.setItem('ShoppingList',
        JSON.stringify({ list: this.list, count: this.count }));
    }
    this.dispatchEvent('update', this.list);
  }

  clear() {
    this.list = [];
    this.save();
  }
}

const service = new ShoppingListServiceSingleton();
export const ShoppingListService = service;
