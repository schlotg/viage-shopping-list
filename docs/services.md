# Viage Shopping List Tutorial

## Services
In Viage, Services are used to manage data or any other global resource. Components should only contain other components and the glu-logic to display data to the user. In contrast, Services should not contain any display logic and handle all aspects of managing the data. This includes notifying components when the data has changed. Services are singletons and do need to be instantiated, you can use them directly by importing them.

### Create a Service
Using the Viage CLI, create a new service by typing the following in a shell in your project's direrctory:

``` viage service shopping-list-service```

This will create a new file at src/services/shopping-list-service.ts. Open this file and it should look like this:

```Javascript
import { Service } from 'viage';

class ShoppingListServiceSingleton extends Service {
  constructor() {
    super();
  }
}
const service = new ShoppingListServiceSingleton ();
export const ShoppingListService = service;
```

Services extend the base class Service, and then export a single instance of the service. To use it in a component you just need to import it:

```Javascript
import { ShoppingListService } from '../service/shopping-list-service';
```

### Defining a Data interface
Let's start by defining an interface for the data. Insert the following lines just above the class declaration:

```Javascript
import { Service } from 'viage';

export interface Item {
  name: string;
  description: string;
  quantity: number;
  purchased: boolean;
  _id?: string; // added by the service
}

class ShoppingListServiceSingleton extends Service {
```

Each Shopping list item is going to have a name, description, quantity, and a purchased flag to indicate if it has been purchased or not. The _id field will be used internally by the service to uniquely track each item. We export the interface so it can be used by components interacting with the data.

### Data Retrieval and Storage
Normally a service would save and retrieve data from a server. Viage doesn't try to duplicate functionality that is already in the DOM. I reccomend using the [fetch API](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch). It is an excellant modern interface that supports promises and is available across the 4 big browsers.

Since writing server code is beyond the scope of this tutorial, we will write to the browser's local storage instead. Add the following code to the shopping-list-service:

```Javascript
class ShoppingListServiceSingleton extends Service {

  private list: Item[] = [];
  private count = 1;

  private generateId() {
    this.count += 1;
    return this.count.toString(16);
  }

  constructor() {
    super();
    if (window.localStorage) {
      const json = window.localStorage.getItem('ShoppingList');
      if (json) {
        const obj = JSON.parse(json) || {};
        this.count = obj.count || 0;
        const list = obj.list || [];
        list.forEach((e: Item) => this.addItem(e));
      }
    }
  }

  save() {
    if (window.localStorage) {
      window.localStorage.setItem('ShoppingList',
        JSON.stringify({ list: this.list, count: this.count }));
    }
    this.dispatchEvent('update', this.list);
  }
```

Our service now contains a local list of the data in a variable named *list*. There is also a local variable called *count* that the *generateId()* function uses to generate unique ids. The *contructor()* and the *save()* function contain code that stores and retrieves data from local storage.

Notice that everytime the *save()* function is called, we dispatch an event called *update*, that lets everyone that is listening know that the data has changed. The *dispatchEvent()* function is defined in the *Service* base class along with the ability to add and remove event listeners. More on that later.

### Getting and Setting Data
Next, we will add some getters and setters to make it easy to add, get, and remove data for the components. This is just standard Typescript. Add the following lines to our service:

```Javascript
  private findIndex(id: string): number {
    let i = 0;
    const list = this.list;
    for (; i < list.length; ++i) {
      if (list[i]._id === id) { break; }
    }
    return i;
  }

  addItem(item: Item) {
    item._id = item._id || this.generateId();
    this.list.push(item);
    this.save();
  }

  removeItem(id: string) {
    const i = this.findIndex(id);
    this.list.splice(i, 1);
    this.save();
  }

  getItem(id: string): Item {
    const i = this.findIndex(id);
    return this.list[i];
  }

  forEach(cb: (item: Item, i?: number) => void) {
    this.list.forEach(cb);
  }

  clear() {
    this.list = [];
    this.save();
  }
```

The function *findIndex* is just a helper function to find an element by id so that we can remove and get items. There is also a *forEach()* function that makes it easy to iterate across all of the items in our list.

Whew! That is the biggest file in our project but it actually does a lot and is great foundation for the rest of the app. All we need to do now is bolt up some components to display the data.

The final file should look like this:

```Javascript
import { Service } from 'viage';

export interface Item {
  name: string;
  description: string;
  quantity: number;
  purchased: boolean;
  _id?: string; // added by the service
}

class ShoppingListServiceSingleton extends Service {

  private list: Item[] = [];
  private count = 1;

  private generateId() {
    this.count += 1;
    return this.count.toString(16);
  }

  constructor() {
    super();
    if (window.localStorage) {
      const json = window.localStorage.getItem('ShoppingList');
      if (json) {
        const obj = JSON.parse(json) || {};
        this.count = obj.count || 0;
        const list = obj.list || [];
        list.forEach((e: Item) => this.addItem(e));
      }
    }
  }

  save() {
    if (window.localStorage) {
      window.localStorage.setItem('ShoppingList',
        JSON.stringify({ list: this.list, count: this.count }));
    }
    this.dispatchEvent('update', this.list);
  }

  private findIndex(id: string): number {
    let i = 0;
    const list = this.list;
    for (; i < list.length; ++i) {
      if (list[i]._id === id) { break; }
    }
    return i;
  }

  addItem(item: Item) {
    item._id = item._id || this.generateId();
    this.list.push(item);
    this.save();
  }

  removeItem(id: string) {
    const i = this.findIndex(id);
    this.list.splice(i, 1);
    this.save();
  }

  getItem(id: string): Item {
    const i = this.findIndex(id);
    return this.list[i];
  }

  forEach(cb: (item: Item, i?: number) => void) {
    this.list.forEach(cb);
  }

  clear() {
    this.list = [];
    this.save();
  }
}

const service = new ShoppingListServiceSingleton();
export const ShoppingListService = service;

```

### [Next (Adding a Shopping List Element Component)](shopping-list-element.md)