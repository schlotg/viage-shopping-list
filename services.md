# Viage Shopping List Tutorial

## Services
In Viage Services are used to manage data or any other global resource. Components should only contain other components and the glu-logic to display data to the user. In contrast Services should not contain and display logic and handle all aspects of managing the data that is being displayed including notifying components when data has changed. Services are singletons and do need to be instantiated, you cna use them directly by importing them.

### Create a Service
Using the Viage CLI, create a new service by typing the following in a shell in your project's direrctory:

``` viage service shopping-list-service```

This will create a new file at src/services/shopping-list-service.ts. Open this file and it should look like this:

```Javascript
import { Service } from 'viage';

class ShoppingListServiceSingleton extends Service {
  constructor() {
    super();
  }
}
const service = new ShoppingListServiceSingleton ();
export const ShoppingListService = service;
```

Services extend the base class Service, and then exports a single instance of the service. To use it in a component you just need to import it:

```Javascript
import { ShoppingListService } from '../service/shopping-list-service';
```

### Defining a Data interface
Let's start by defining an interface for the data. Insert the following lines just above the class declaration:

```Javascript
import { Service } from 'viage';

export interface Item {
  name: string;
  description: string;
  quantity: number;
  purchased: boolean;
  _id?: string; // added by the service
}

class ShoppingListServiceSingleton extends Service {
```

Each Shopping list item is going to have a name, description, quantity, and a purchased flag to indicated if purchased or not. The _id field will be used internally by the service to uniquely track each item. We export the interface so it can be used by components interacting with the data.



