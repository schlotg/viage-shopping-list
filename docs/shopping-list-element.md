# Viage Shopping List Tutorial

## Adding a Shopping List Element Component
Now we need a component to display a shopping list item.

### Generate a New Component
Use the CLI to generate a new component by typing:

```viage component shopping-list-element```

This will generate a empty component at src/components/shopping-list-element.ts

### Modify the Component

Modify the newly generated component so it looks like this:

```Javascript
import { Component } from 'viage';
import { Item, ShoppingListService } from '../services/shopping-list-service';
import { getRouter } from 'viage';

export class ShoppingListElement extends Component {

  private item: Item;

  constructor(){
    super('shopping-list-element');
  }

  init(item: Item) {
    this.item = item;
    this.setHTML(`
      <div style="width: 100%; margin-top: 5px; margin-bottom: 5px;">
        <input attach="purchased" type="checkbox" ${item.purchased ? "checked": ""} />
        <span style="width: 30px; display: inline-block">${item.quantity}</span>
        <span style="width: 100px; display: inline-block">${item.name}</span>
        <span style="width: 300px; display: inline-block">${item.description}</span>
        <button attach="delete">Delete</button>
        <button attach="edit">Edit</button>
      </div>
    `);
    this.attachments.delete.addEventListener('click', () => {
      ShoppingListService.removeItem(this.item._id);
    });
    const purchased = this.attachments.enabled;
    purchased.addEventListener('click', () => {
      this.item.purchased = purchased.checked;
      ShoppingListService.save();
    });
    this.attachments.edit.addEventListener('click', () => {
      getRouter('main').go(`#edit/${this.item._id}`);
    });
    return this;
  }
}
```

### Data
You will notice that the component has a member called *item*. This is a local copy of the data. Every time *init()* is called with a Item, the local copy is set and the component's HTML is replaced with new HTML that represents the new state. We use the back-tick strings template functionality to inject the various item values into the HTML.

### Attaching
There are two buttons in the component that need click handlers attached to them. Viage has a convienent way of accessing HTML elements within a component. When you specify *attach=name* in the HTML it will automatically be able to be accessed in the *this.attachments[name]* . We utilize this feature to addEventListeners to both the buttons and the purchased checkbox.

You can see that the delete button calls the *ShoppingListService.removeItem()* function. The purchased click handler triggers a save in the *ShoppingListService*. The edit button triggers a new Router state to edit the item. We will cover the Router in one of the later sections.

Notice that we are using the standard DOM API addEventListener() instead of re-implementing it.

### [Next (The Shopping List Component)](docs/shopping-list.md)
