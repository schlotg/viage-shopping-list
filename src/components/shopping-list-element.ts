import { Component } from 'viage';
import { Item, ShoppingListService } from '../services/shopping-list-service';
import { States } from './app';

interface Id {
  id: string;
}

export class ShoppingListElement extends Component {

  private item: Item;

  constructor() {
    super('shopping-list-element');
  }

  init(item: Item) {
    this.item = item;
    this.setHTML(`
      <div class="not-mobile list-item">
        <input attach="enabled" type="checkbox" ${item.purchased ? "checked": ""} />
        <span style="width: 30px; display: inline-block">${item.quantity}</span>
        <span style="width: 20%; display: inline-block">${item.name}</span>
        <span style="width: 40%; display: inline-block">${item.description}</span>
        <button attach="delete" style="padding: 1px; float: right">Delete</button>
        <button attach="edit" style="padding: 1px; float: right; margin-right: 10px;">Edit</button>
      </div>
      <div attach="m_edit" class="mobile list-item">
        <input attach="m_enabled" type="checkbox" ${item.purchased ? "checked": ""} />
        <span style="width: 30px; display: inline-block">${item.quantity}</span>
        <span style="width: 50%; display: inline-block">${item.name}</span>
        <button attach="m_delete" style="padding: 1px; float: right">X</button>
        <div style="display: block; margin-top:20px;">
          <label style="font-weight:800">Description: </label>
          <span style="width: 40%;" >${item.description}</span>
        </div>
      </div>

    `);
    const attachments = this.attachments;
    const remove = () => ShoppingListService.removeItem(this.item._id);
    const enable1 = () => {
      this.item.purchased = attachments.enabled.checked;
      ShoppingListService.save();
    };
    const enable2 = () => {
      this.item.purchased = attachments.m_enabled.checked;
      ShoppingListService.save();
    };
    const edit = () => {
      const url = this.router.createUrl<Id>(States.EDIT, {id: this.item._id});
      this.router.go(url);
    };
    attachments.delete.addEventListener('click', remove);
    attachments.m_delete.addEventListener('click', remove);
    attachments.enabled.addEventListener('click', enable1);
    attachments.m_enabled.addEventListener('click', enable2);
    this.attachments.edit.addEventListener('click', edit);
    this.attachments.m_edit.addEventListener('click', edit);
    return this;
  }
}