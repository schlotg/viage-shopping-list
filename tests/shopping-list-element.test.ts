import { Item, ShoppingListService } from '../src/services/shopping-list-service';
import { ShoppingListElement } from '../src/components/shopping-list-element';
import { createRouter, destroyRouter } from 'viage';
import { item1 as item, triggerClickEvent } from './common';

test('Shopping List Element should initialize correctly with test data', () => {
  const sle = new ShoppingListElement();
  sle.init(item);
  const html = sle.e.innerHTML;
  expect(html).toMatchSnapshot();
});

test('Shopping List Element delete handler should work', () => {
  const sle = new ShoppingListElement();
  sle.init(item);
  const mock = jest.spyOn(ShoppingListService, 'removeItem');
  const deleteElement = sle.e.querySelector('[attach=delete]') as HTMLElement;
  triggerClickEvent(deleteElement);
  expect(mock).toBeCalled();
  mock.mockRestore();
});

test('Shopping List Element edit handler should work', () => {
  const router = createRouter('test', document.createElement('div'), 'LOCATION' );
  const sle = new ShoppingListElement();
  sle.init(item);
  const mock = jest.spyOn(router, 'go');
  const mock2 = jest.spyOn(router, 'createUrl');
  sle.setRouter(router);
  const editElement = sle.e.querySelector('[attach=edit]') as HTMLElement;
  triggerClickEvent(editElement);
  expect(mock).toBeCalled();
  mock.mockRestore();
  mock2.mockRestore();
  destroyRouter('test');
});

test('Shopping List Element enabled handler should work', () => {
  const sle = new ShoppingListElement();
  sle.init(item);
  const mock = jest.spyOn(ShoppingListService, 'save');
  const element = sle.e.querySelector('[attach=enabled]') as HTMLInputElement;
  triggerClickEvent(element);
  expect(mock).toBeCalled();
  expect(element.checked).toBe(true);
  mock.mockRestore();
});
