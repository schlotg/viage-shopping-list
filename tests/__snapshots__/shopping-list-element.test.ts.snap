// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shopping List Element should initialize correctly with test data 1`] = `
"
      <div style=\\"width: 100%; margin-top: 5px; margin-bottom: 5px;\\">
        <input attach=\\"enabled\\" type=\\"checkbox\\">
        <span style=\\"width: 30px; display: inline-block\\">1</span>
        <span style=\\"width: 100px; display: inline-block\\">Dog Food</span>
        <span style=\\"width: 300px; display: inline-block\\">30 lb bag</span>
        <button attach=\\"delete\\" style=\\"padding: 1px;\\">Delete</button>
        <button attach=\\"edit\\" style=\\"padding: 1px;\\">Edit</button>
      </div>
    "
`;
